File: src/main/java/info/idgst/digest/DigestResource.java
package info.idgst.digest;

import info.idgst.exception.DigestAlreadyExistsException;
import info.idgst.exception.RestException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;

/**
 * Endpoint for Digest.
 *
 * @author Aliaksei Bahdanau
 */
@RestController
@RequestMapping(value = "/api/v1/digests")
public class DigestResource {

    private final DigestService digestService;

    @Autowired
    public DigestResource(final DigestService digestService) {
        this.digestService = digestService;
    }

    @RequestMapping(method = RequestMethod.GET)
    public ResponseEntity<Page<Digest>> digests(
            @RequestParam(value = "page", required = false, defaultValue = "0") int page,
            @RequestParam(value = "size", required = false, defaultValue = "10") int size,
            @RequestParam(value = "sortDirection", required = false, defaultValue = "DESC") String sortDirection,
            @RequestParam(value = "sortBy", required = false, defaultValue = "publishedDate") String sortBy) {
        Page<Digest> digests = digestService.findAll(page, size, Sort.Direction.fromString(sortDirection), sortBy);
        if (!digests.hasContent()) {
            return new ResponseEntity<>(digests, HttpStatus.NOT_FOUND);
        }
        return new ResponseEntity<>(digests, HttpStatus.OK);
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public ResponseEntity<Digest> findById(@PathVariable("id") String id) {
        Digest digest = digestService.findById(id);
        if (digest == null) {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }

        return new ResponseEntity<>(digest, HttpStatus.OK);
    }

    @RequestMapping(method = RequestMethod.POST, consumes = "application/json")
    public ResponseEntity<Digest> save(@RequestBody Digest digest) {
        Digest savedDigest;
        try {
            savedDigest = digestService.save(digest);
        } catch (DigestAlreadyExistsException e) {
            throw new RestException(0, "Error while saving Digest", e.getMessage());
        }

        URI pollURI = ServletUriComponentsBuilder.fromCurrentRequest()
                                                 .path("/{id}")
                                                 .buildAndExpand(savedDigest.getId())
                                                 .toUri();
        HttpHeaders responseHeaders = new HttpHeaders();
        responseHeaders.setLocation(pollURI);

        return new ResponseEntity<>(savedDigest, responseHeaders, HttpStatus.CREATED);
    }
}

File: src/main/kotlin/info/devbug/digest/DigestResource.kt
package info.devbug.digest

import info.devbug.api.RestException
import info.devbug.digest.repository.DigestDto
import info.devbug.digest.service.DigestService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.support.ServletUriComponentsBuilder
import java.net.URI

/**
 * Endpoint for Digest.
 *
 * @author Aliaksei Bahdanau
 */
@RestController
@RequestMapping(value = "/api/v1/digests")
class DigestResource @Autowired constructor(val digestService: DigestService) {

    /**
     * Retrieves all [DigestDto]s with pagination support.
     *
     * @param page to be returned.
     * @param size number of items to be returned.
     * @return [ResponseEntity]
     */
    @RequestMapping(method = arrayOf(RequestMethod.GET))
    fun digests(@RequestParam(value = "page", required = false, defaultValue = "0") page: Int,
                @RequestParam(value = "size", required = false, defaultValue = "10") size: Int):
            ResponseEntity<Page<DigestDto>> {
        val digests = digestService.findAll(page, size)
        return ResponseEntity(digests, HttpStatus.OK)
    }

    /**
     * Saves the specified digest.
     *
     * @param digest [DigestDto] for saving.
     * @throws Exception in case when Digest saving failed.
     * @return [ResponseEntity] with saved Digest and status '201 Created' in case when Digest is saved successfully.
     */
    @RequestMapping(method = arrayOf(RequestMethod.POST), consumes = arrayOf("application/json"))
    fun save(@RequestBody digest: DigestDto): ResponseEntity<DigestDto> {
        val savedDigest: DigestDto
        try {
            savedDigest = digestService.save(digest)
        } catch(e: Exception) {
            throw RestException(0, "Error while saving Digest",
                    "Digest with such title already exists.")
        }

        val responseHeaders: HttpHeaders = HttpHeaders()
        val newPollUri: URI = ServletUriComponentsBuilder
                .fromCurrentRequest()
                .path("/{id}")
                .buildAndExpand(savedDigest.id)
                .toUri()
        responseHeaders.location = newPollUri

        return ResponseEntity(savedDigest, responseHeaders, HttpStatus.CREATED)
    }

    @RequestMapping(value = "/{id}", method = arrayOf(RequestMethod.GET))
    fun findById(@PathVariable("id") id: String): ResponseEntity<DigestDto> {
        val digest = digestService.findById(id) ?: return ResponseEntity(HttpStatus.NOT_FOUND)
        return ResponseEntity(digest, HttpStatus.OK)
    }
}